### PR TITLE
[Internal] OpenTelemetry: Adds diagnostics for network level requests

### LINKED ISSUE #3459: [Internal] Designing capabilities to support Distributed Tracing
**Is your feature request related to a problem? Please describe.**
To be able to implement and deliver distributed tracing transactions downstream of the Cosmos DB SDK into the compute gateway, routing gateway, and/or the backend services.

**Describe the solution you'd like**
While conforming to the  OpenTelemetry Guidelines, Standards and Best Practices, we will build the following capabilities:

- The ability to generate a new _TraceId_, as well as additional intrinsic and extrinsic context information, that represents a new trace transaction from the SDK client automatically, when one is not provided by the user.

- The ability to create accept a _TraceId_, as well as additional intrinsic and extrinsic contextual information, that represents a new or existing trace transaction from the SDK client manually, when one is provided by the user.

- The ability to transport a generated or manually created _TraceId_, as well as additional intrinsic and extrinsic contextual information, from the SDK into other components that live within that timeline downstream of the SDK.

- The ability to generate _SpanId_, as well as additional intrinsic and extrinsic contextual information, that represents a new component, or operation within that trace's transaction.

**Describe alternatives you've considered**
TBD

**Additional context**
TBD

### PR DESCRIPTION
**Pull request by aavasthy**
# Pull Request Template

## Description

This PR adds open telemetry support to the SDK for network level requests for both HTTP and TCP requests.
It adds Operation level ActivityId(Parent id for network level activities) as part of diagnostics log.

This is a draft PR and would have the following once updated Direct package is included:
- Direct package version update
- Emulator Tests for E2E flow
- IsEnableDistributedTracing flag being passed to StoreClientFactory

## Type of change

Please delete options that are not relevant.

- [] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #3459

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs [hunk 1]
```diff
@@ -501,15 +503,25 @@ protected virtual void Dispose(bool disposing)
             {
                 try
                 {
-                    TResult result = await task(trace).ConfigureAwait(false);
-                    if (openTelemetry != null && recorder.IsEnabled)
+                    if (recorder.IsEnabled == false && this.clientOptions.IsDistributedTracingEnabled)
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:506]: this.clientOptions.IsDistributedTracingEnabled should be !this.clientOptions.IsDistributedTracingEnabled

File: Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs [hunk 2]
```diff
@@ -93,7 +93,7 @@ public TransportHandler(CosmosClient client)
 
             DocumentServiceRequest serviceRequest = request.ToDocumentServiceRequest();
 
-            ClientSideRequestStatisticsTraceDatum clientSideRequestStatisticsTraceDatum = new ClientSideRequestStatisticsTraceDatum(DateTime.UtcNow, request.Trace.Summary);
+            ClientSideRequestStatisticsTraceDatum clientSideRequestStatisticsTraceDatum = new ClientSideRequestStatisticsTraceDatum(DateTime.UtcNow, request.Trace.Summary, Activity.Current?.Id);
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs:96]: why do you need to pass this? its static variable and will be available everywhere.

File: Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs [hunk 3]
```diff
@@ -194,6 +194,11 @@ public void Visit(ClientSideRequestStatisticsTraceDatum clientSideRequestStatist
                 this.jsonWriter.WriteFieldName("Id");
                 this.jsonWriter.WriteStringValue("AggregatedClientSideRequestStatistics");
 
+                if (!String.IsNullOrEmpty(clientSideRequestStatisticsTraceDatum.TraceId))
+                {
+                    this.jsonWriter.WriteFieldName("TraceId");
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceWriter.TraceJsonWriter.cs:199]: some better name for TraceId? may be _DistributedTacingId_

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs (hunk 1 above)
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:506]: ideally you should create one more type of _OpenTelemetryCoreRecorder_ which will create and dispose the activity for you. and recording attributes should be noop.

File: Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs (hunk 2 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs]: updated

**Pull request closed without merging**